drivers: input: cap1203, support 3 to 8 channels

lfelten opened the pull request:
The Microchip CAP12xx series are available in 3, 6 or 8 channel versions.
This requires existing device trees to add the node property 'channels'.

This patch adds support for CAP1206 or CAP1208 in addition to CAP1203. 
Support for a CAP1214 14 channel version can be added but requires multiple changes to the driver, mostly because the fourteen bits are spread across two registers.

The driver supports polling mode and interrupt mode.
The driver was tested with a TTGO T8S3 and a CAP1206 in interrupt and polling mode.

I have two spare test PCBs with CAP1206 available, if someone is interested.

fabiobaltieri commented on dts/bindings/input/microchip,cap12xx.yaml:
```diff
@@ -0,0 +1,29 @@
+      Array of input event key codes (INPUT_KEY_* or INPUT_BTN_*).
+
+  repeat:
+    type: boolean
```
would it make sense to have a default value here?

lfelten replied:
Thanks for the review.
This is also related to your next comment, the default is set in the C file and defaults to false.
What is the preferred method?
My reasoning was: you don't specify "repeat" - it will not repeat. You add "repeat" - it repeats

fabiobaltieri replied:
sorry I was meaning to delete this one - booleans default to false implicitly, my comment was wrong :-)

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,232 @@
+		.int_gpio = GPIO_DT_SPEC_INST_GET_OR(index, int_gpios, {0}),                       \
+		.input_channels = cap12xx_input_channels,                                          \
+		.input_codes = cap12xx_input_codes_##inst,                                         \
+		.repeat = DT_INST_PROP_OR(index, repeat, 0),                                       \
```
booleans don't need the `_OR` variants, just `DT_INST_PROP`

lfelten replied:
Then I need to set a default in the device tree.
(Need to look up how to do that)

fabiobaltieri replied:
you shouldn't have to, the boolean defaults to false, just change to `DT_INST_PROP` and it'll work as before

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,232 @@
+
+#define CAP12XX_INIT(index)                                                                        \
+	BUILD_ASSERT(DT_INST_PROP_LEN(index, input_codes) == DT_INST_PROP(index, channels));       \
+	static const uint8_t cap12xx_input_channels = DT_INST_PROP(index, channels);               \
```
you probably want to use `cap12xx_input_channels_##inst` as variable name or this would break if you had multiple instances

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,232 @@
+			return r;
+		}
+	}
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+	else {
```
confused, what if you don't specify the interrupt gpio and don't enable `CONFIG_INPUT_CAP12XX_POLL` either? Nothing would work right?

Either way bring the `ifdef` in the block, not a fan of splitting the `} else {` like that

lfelten replied:
If you don't use polling, it's interrupts. If you don't specify an interrupt gpio, you get an error message.
But it will indeed not work.
I can change the code structure, this is based on the previous driver structure.

fabiobaltieri replied:
> If you don't specify an interrupt gpio, you get an error message

Am I missing something? If `config->int_gpio.port` is NULL and `CONFIG_INPUT_CAP12XX_POLL=n` this would fall through and initialize with no way to work, right?

lfelten replied:
No, you're right. If you do nothing at all, it will just add driver code. That's what the old driver did too.
What is the preferred behavior?
Should I LOG_ERR at run time?
There probably is a way to check if there is a device tree node at compile time too, you happen have an example of this?
Maybe I'll just #ifdef / #elseif the poll vs int parts depending on CONFIG_INPUT_CAP12XX_POLL.
Let me rewrite some more of that driver. 
(My intent was to just do a minimal change :-) )

fabiobaltieri replied:
>  (My intent was to just do a minimal change :-) )

I  know sorry, told you you should have had a separate rename commit :-)

Two options:
- bring the ifdef into the else statement and if it's not set return an error
```c
} else {
#if CONFIG...POLL
        /* setup poll */
#else
        LOG_ERR("CONFIG_POLL must be enabled etc etc");
        return -EINVAL;
#endif

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,232 @@
+			      K_MSEC(CONFIG_INPUT_CAP12XX_PERIOD));
+	}
+#endif
+	if (config->repeat == 0) {
```
`if (config->repeat) {`

it's a boolean...

lfelten replied:
booleans in C are suspicious :-)
I'll fix that.

fabiobaltieri replied:
yeah to be honest I tend to deny the existance of booleans in C as well, but that's the style used in the code base :-)

fabiobaltieri reviewed (commented):
sorry I probably commented on the existing implementation, would have been a better idea to split the PR in two commits, one just the rename, one for the changes, just a note for the future, don't bother doing it now, good chance to fix some existing minor issues as well

fabiobaltieri commented:
Also check the CI failure

```
devicetree error: 'channels' is marked as required in 'properties:' in /__w/zephyr/zephyr/dts/bindings/input/microchip,cap12xx.yaml, but does not appear in <Node /test/i2c@1/cap12xx@3 in '/__w/zephyr/zephyr/misc/empty_file.c'>
```

lfelten commented:
@fabiobaltieri 
I addressed the issues you mentioned. This resulted in major changes.
I also converted this PR to draft and would appreciate it if you'd review the 'new' driver.

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,229 @@
+
+	gpio_init_callback(&data->int_gpio_cb, cap12xx_isr_handler, BIT(config->int_gpio.pin));
+
+	r = gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
```
`gpio_add_callback_dt`

fabiobaltieri commented:
Hey could you also add a line in `doc/releases/migration-guide-4.1.rst` mentioning the compatible change and whatever is needed to migrate existing use caes? thanks!

lfelten commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,229 @@
+}
+
+#define CAP12XX_INIT(index)                                                                        \
+	BUILD_ASSERT(DT_INST_PROP_LEN(index, input_codes) == DT_INST_PROP(index, channels));       \
```
The BUILD_ASSERT macro errors are unreadable (at least hard to decode) when no message is added, so I have this:
BUILD_ASSERT(DT_INST_NODE_HAS_PROP(index, int_gpios), "cap12xx: device tree node needs int-gpios property");
It seems to work, needs more testing later.

Adding a dependency to !CONFIG_INPUT_CAP12XX_POLL is hard, because that is not defined when not selected.
I just wonder if I should remove the Kconfig option altogether and move to a device tree only config?
Then I would need a BUILD_ASSERT(DT_INST_NODE_HAS_PROP(index, int_gpios) "and not" DT_INST_NODE_HAS_PROP(index, poll_interval),"cap12xx: use either interrupt or polling mode")
(how do I do "and not" in C-preprocessor statements?)

If I remove the Kconfig polling option, should I define a "CAP12XX_POLL" depending on DT_INST_NODE_HAS_PROP(index, poll_interval) at the top?

With more macro magic, I could remove the "channels" property of the device tree and use the property length of the "input_codes" property as channel count. 

It comes down to this:
Do I replace Kconfig options by device tree properties?

fabiobaltieri replied:
Sorry down to the main question:

> Do I replace Kconfig options by device tree properties?

Yes I think it'd be a better solution, if anything because it means that the CI test build is always going to cover all the code. It's a more convoluted change but if you feel like doing I think it'd lead to a better driver. Otherwise if you can get that build assert sorted I think that'd be good enough. Your choice.

lfelten committed:
drivers: input: cap1203, rename to cap12xx

rename cap1203 to cap12xx to support 3 to 8 channels

Signed-off-by: Lothar Felten <lothar.felten@gmail.com>

lfelten committed:
drivers: input: cap12xx, support 3 to 8 channels

The Microchip CAP12xx series are available in 3, 6 or 8 channel versions.

Co-authored-by: Benjamin Cabé <kartben@gmail.com>

Signed-off-by: Lothar Felten <lothar.felten@gmail.com>

lfelten commented:
@fabiobaltieri 
I tried to replace the Kconfig by checking the device tree for specific properties:
I need some way to define a compile time variable to select between polling and interrupt, and this could be different for each instance (not sure if that is possible to support at compile time, I doubt it).
At the end of the file I added DT_INST_FOREACH_STATUS_OKAY(CAP12XX_CONFIG) but now I'm stuck.

fabiobaltieri commented:
> I need some way to define a compile time variable to select between polling and interrupt, and this could be different for each instance (not sure if that is possible to support at compile time, I doubt it).

Change the config structure to use two pointers for gpio and timer (using the old driver code as an example)

```c
struct cap1203_config {
        struct i2c_dt_spec i2c;
        const uint16_t *input_codes;
        struct gpio_dt_spec *int_gpio; // this
        struct k_timer *timer; // and this
};
```

then in `CAP1203_INIT` initialize only one or the other depending on whether you have the gpio defined or not, something like

```c
        COND_CODE_1(DT_INST_NODE_HAS_PROP(index, int_gpios), (
        static const struct gpio_dt_spec cap1203_int_gpio_##index = GPIO_DT_SPEC_INST_GET(index, int_gpios);
        ), (
        static const struct k_timer cap1203_poll_timer_##index;
        ))

        static const struct cap1203_config cap1203_config_##index = {
                .i2c = I2C_DT_SPEC_INST_GET(index),
                .input_codes = cap1203_input_codes_##inst,
                COND_CODE_1(DT_INST_NODE_HAS_PROP(index, int_gpios), (
                .int_gpio = &cap1203_int_gpio_##index,
                ), (
                .poll_timer = &cap1203_poll_timer_##index,
                )
        };
```

(dropped the `\` for simplicity and I haven't actually built it, may need some massaging but you got the idea)

on the code you have to change few things to make it dereference `int_gpio` and `poll_timer`, and you can check which one is not `NULL` to see in runtime if you are in poll or interrupt mode. At this point there's nothing to build check because if one doe snot specify the gpio the driver goes into poll mode automatically.

lfelten commented:
@fabiobaltieri 
This is still a draft/work in progress. I did some tests with the hardware, but I'll need more time to test more.

fabiobaltieri commented on dts/bindings/input/microchip,cap12xx.yaml:
```diff
@@ -10,6 +10,18 @@ include: i2c-device.yaml
+    description: |
+      Property to enable the interrupt repeat mode for prolonged touch events.
+
+  poll-interval:
```
`poll-interval-ms`, then you can tell the unit without having to refer to the binding doc

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -4,45 +4,45 @@
 	const uint16_t *input_codes;
+	struct gpio_dt_spec *int_gpio;
+	bool repeat;
+	const uint16_t poll_interval;
```
`poll_interval_ms`

fabiobaltieri commented on drivers/input/input_cap12xx.c:
```diff
@@ -136,77 +138,93 @@ static int cap1203_init(const struct device *dev)
-	if (config->int_gpio.port != NULL) {
-		if (!gpio_is_ready_dt(&config->int_gpio)) {
-			LOG_ERR("Interrupt GPIO controller device not ready");
+	if ((config->int_gpio) == NULL) {
```
no need for the innter parenthesis, just `if (config->int_gpio == NULL) {`

fabiobaltieri reviewed (commented):
cool, looks good to me just a couple nitpicks, thanks for going through this

lfelten commented:
@fabiobaltieri 
Hi Fabio, thanks for your feedback, I updated the code and successfully tested interrupt and polling mode.

fabiobaltieri commented:
Yeah that's fine, can you add that to this PR?

lfelten commented:
> Yeah that's fine, can you add that to this PR?

of course

fabiobaltieri reviewed (dismissed):
Great stuff thanks for working on this.

kartben commented on dts/bindings/input/microchip,cap12xx.yaml:
```diff
@@ -0,0 +1,30 @@
+    type: array
+    required: true
+    description: |
+      array of input event key codes (INPUT_KEY_* or INPUT_BTN_*).
```
```suggestion
      Array of input event key codes (INPUT_KEY_* or INPUT_BTN_*).
```

kartben commented on dts/bindings/input/microchip,cap12xx.yaml:
```diff
@@ -0,0 +1,30 @@
+  int-gpios:
+    type: phandle-array
+    description: |
+      interrupt gpio when not using polling mode.
```
```suggestion
      Interrupt GPIO when not using polling mode.
```

kartben commented on dts/bindings/input/microchip,cap12xx.yaml:
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2022 Keiya Nobuta
+# SPDX-License-Identifier: Apache-2.0
+
+description: cap12xx 3, 6 or 8-channel capacitive touch sensor
```
```suggestion
description: Microchip CAP12xx family of 3, 6 and 8-channel capacitive touch sensors.
```

kartben commented on drivers/input/Kconfig.cap12xx:
```diff
@@ -0,0 +1,11 @@
+	depends on DT_HAS_MICROCHIP_CAP12XX_ENABLED
+	select I2C
+	help
+	  Enable driver for microchip cap12xx multi-channel capacitive
```
```suggestion
	  Enable driver for Microchip CAP12xx multi-channel capacitive
```

kartben commented on drivers/input/Kconfig.cap12xx:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config INPUT_CAP12XX
+	bool "cap12xx multi-channel capacitive touch sensor driver"
```
```suggestion
	bool "Microchip CAP12xx multi-channel capacitive touch sensor driver"
```

kartben reviewed (changes requested):
Some documentation comments, and a question:
I wonder if it was really necessary to drop the cap1203 compatible? It feels like it would be less intrusive for downstream users if the old binding would still exist, and just tap into the new generic driver? something similar to this construct? https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/ti/ina23x/ina230.c#L273-L280 cc @fabiobaltieri

kartben commented on drivers/input/input_cap12xx.c:
```diff
@@ -0,0 +1,230 @@
+			LOG_DBG("cap12xx disabled repeated interrupts");
+		}
+	}
+	LOG_DBG("cap12xx: %d channels configured", config->input_channels);
```
no need for the cap12xx prefix here afaict

```suggestion
	LOG_DBG("%d channels configured", config->input_channels);
```

fabiobaltieri commented:
Can you squash the generated fixup commits in the last one? We can't disable that button on GitHub unfortunately

lfelten commented:
> Can you squash the generated fixup commits in the last one? We can't disable that button on GitHub unfortunately

@fabiobaltieri just did that and also rebased it to main.
